java-support/tck/src/main/java/io/cloudstate/javasupport/tck/model/action/ActionTckModelBehavior.java

```diff
@@ -0,0 +1,111 @@
+import java.util.concurrent.CompletionStage;
+
+@Action
+public class ActionTckModelBehavior {
+
+  private final ActorSystem system = ActorSystem.create("ActionTckModel");
+
+  public ActionTckModelBehavior() {}
+
+  @CallHandler
+  public CompletionStage<ActionReply<Response>> processUnary(
+      Request request, ActionContext context) {
+    return Source.single(request).via(responses(context)).runWith(singleResponse(), system);
+  }
+
+  @CallHandler
+  public CompletionStage<ActionReply<Response>> processStreamedIn(
+      Source<Request, NotUsed> requests, ActionContext context) {
+    return requests.via(responses(context)).runWith(singleResponse(), system);
+  }
+
+  @CallHandler
+  public Source<ActionReply<Response>, NotUsed> processStreamedOut(
+      Request request, ActionContext context) {
+    return Source.single(request).via(responses(context));
+  }
+
+  @CallHandler
+  public Source<ActionReply<Response>, NotUsed> processStreamed(
+      Source<Request, NotUsed> requests, ActionContext context) {
+    return requests.via(responses(context));
+  }
+
+  private Flow<Request, ActionReply<Response>, NotUsed> responses(ActionContext context) {
+    return Flow.of(Request.class)
+        .flatMapConcat(request -> Source.from(request.getGroupsList()))
+        .map(group -> response(group, context));
+  }
+
+  private ActionReply<Response> response(ProcessGroup group, ActionContext context) {
+    ActionReply<Response> reply = ActionReply.noReply();
+    for (ProcessStep step : group.getStepsList()) {
+      switch (step.getStepCase()) {
+        case REPLY:
+          reply =
+              ActionReply.message(
+                      Response.newBuilder().setMessage(step.getReply().getMessage()).build())
+                  .withEffects(reply.effects());
+          break;
+        case FORWARD:
+          reply =
+              ActionReply.<Response>forward(serviceTwoRequest(context, step.getForward().getId()))
+                  .withEffects(reply.effects());
+          break;
+        case EFFECT:
+          SideEffect effect = step.getEffect();
+          reply =
+              reply.withEffects(
+                  Effect.of(serviceTwoRequest(context, effect.getId()), effect.getSynchronous()));
+          break;
+        case FAIL:
+          reply =
+              ActionReply.<Response>failure(step.getFail().getMessage())
+                  .withEffects(reply.effects());
+      }
+    }
+    return reply;
+  }
+
+  private Sink<ActionReply<Response>, CompletionStage<ActionReply<Response>>> singleResponse() {
+    return Sink.fold(
+        ActionReply.noReply(),
+        (reply, next) ->
+            next.isEmpty() ? reply.withEffects(next.effects()) : next.withEffects(reply.effects()));
+  }
+
+  private ServiceCall serviceTwoRequest(Context context, String id) {
+    return context
+        .serviceCallFactory()
+        .lookup("cloudstate.tck.model.action.ActionTwo", "Call", OtherRequest.class)
```

Comment:
No issue here, but I'm surprised to see the Java gRPC service descriptors same as, the ones of the Go prototbuf plugin, do not export typesafe identifiers for services and service names.

Reply:
… there is work underway to change that for the go gRPC plugin https://github.com/grpc/grpc-go/issues/3762

Reply:
The Scala service descriptors do have the identifiers. Maybe they've been added for Java as well now. Good that there will be a better API for Go.

Reply:
Akka gRPC for Java does of course generate a bunch of things, including static service names that can be used here. The method descriptors in the client all look to be private, otherwise they could be used for a lookup here. Agree that it would be good to have typesafe APIs for Cloudstate service calls, and we should otherwise look at generating these. The Java support is very reflection and runtime-based currently.

Reply:
I've updated this to use the generated service name at least, from Akka gRPC (the generation was actually disabled, as only the java protoc generation is needed).